Subject: dasd: move wake_up call
From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Patch-mainline: v3.6-rc7
Git-commit: 1f08be80bef6f7a3faaa728db836b47ff742f41f
References: bnc#786976,LTC#86252

Symptom:      Tools like dasdfmt or fdasd complain about "disk in use" when
              there should be no user of the dasd device.

              http://www.mail-archive.com/linux-390@vm.marist.edu/msg62153.html

Problem:      During a dasd state change its possible that udev is notified
              after the process waiting for this state change is woken up.
              This could lead to udev still examining the dasd block device
              even after a call to udevadm settle returned.
Solution:     Ensure that all work is done when the process waiting for a
              dasd state change is woken up. With this change it is save
              to assume that after a userspace triggered state change and
              a udev settle invocation there are no unexpected users of a
              dasd device.

              Upstream via:
              1f08be8 s390/dasd: move wake_up call
Reproduction: -

Acked-by: John Jolly <jjolly@suse.de>
--- a/drivers/s390/block/dasd.c
+++ b/drivers/s390/block/dasd.c
@@ -497,11 +497,11 @@ static void dasd_change_state(struct das
 	if (rc)
 		device->target = device->state;
 
-	if (device->state == device->target)
-		wake_up(&dasd_init_waitq);
-
 	/* let user-space know that the device status changed */
 	kobject_uevent(&device->cdev->dev.kobj, KOBJ_CHANGE);
+
+	if (device->state == device->target)
+		wake_up(&dasd_init_waitq);
 }
 
 /*
